[#109030254] Terraform container + README

combor opened the pull request:
# What

Story: https://www.pivotaltracker.com/n/projects/1275640/stories/109030254

Terraform (0.6.7) container with instructions how to
build and run it.
# How to test

Build it:
- `git clone https://github.com/government-paas-docker/docker-terraform.git`
- `cd docker-terraform`
- `docker build -t terraform .`

Run it:

```
docker run -ti \
        -e TF_VAR_AWS_ACCESS_KEY_ID=$TF_VAR_AWS_ACCESS_KEY_ID \
        -e TF_VAR_AWS_SECRET_ACCESS_KEY=$TF_VAR_AWS_SECRET_ACCESS_KEY \
        terraform
```

Check whether you can run terraform and variables are exported.
# Who can test

anyone but @combor

Commit:
Terraform container + README

Initial readme and terraform (0.6.7) container with instructions how to
build and run it.

List of packages:

terraform - downloaded from Hashicorp
wget - downloads terraform
unzip - unzips terraform
ssh - terraform uses it to upload files
ruby - bosh

dcarley commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+ENV TERRAFORM_VER 0.6.7
+ENV TERRAFORM_ZIP terraform_"$TERRAFORM_VER"_linux_amd64.zip
+
+RUN apt-get update && apt-get -y install wget unzip openssh-client dnsutils git pass ruby
```
Do we need `pass` here?

Can you include some detail in the commit message about why the other packages are listed. I can figure it out - but it would be good to be explicit.

combor replied:
> Do we need pass here?

It depends how we want to pass CF credentials. 

Other packages:
- wget - downloads terraform
- unzip - unzips terraform
- ssh - terraform uses it to upload files
- dnsutils - debugging (nslookup/dig etc.)
- git - clone pass repo
- ruby - bosh

dcarley replied:
CF is out of scope for now - we're only provisioning Concourse to start with. Either way we should keep the container as minimal as possible and pass (no pun intended!) the credentials as an input.

Thanks for the list. Could you add it to the commit message for posterity?

Similarly do we need dnsutils? If we were debugging in a container then we can add those in at runtime.

combor replied:
Ok, fair enough. Pass, git, dnsutils are removed now and commit message modified.

dcarley commented on README.md:
````diff
@@ -1 +1,18 @@
+```
+docker run -ti \
+        -e TF_VAR_AWS_ACCESS_KEY_ID=$TF_VAR_AWS_ACCESS_KEY_ID \
+        -e TF_VAR_AWS_SECRET_ACCESS_KEY=$TF_VAR_AWS_SECRET_ACCESS_KEY \
````
I think the right-hand side of these variables should not be prefixed by `TF_VAR_`. They should be the standard environment variables `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`, that you would have exported on your host machine.

combor replied:
Well, right hand side can be anything. That's just an example. I will correct it if it looks better with `AWS_ACCESS_KEY_ID`

dcarley replied:
Yes please

dcarley commented on README.md:
````diff
@@ -1 +1,18 @@
+
+## How to run and build
+
+- ```git clone https://github.com/combor/docker-terraform.git```
````
This refers to your own account. I think we could probably omit the clone/cd instructions altogether.

combor replied:
Yeah, copy paste error. Fixed

dcarley commented:
Thanks for the updates. I got a build error when building locally against boot2docker. Looks like a problem with variable interpolation:
```
Step 5 : RUN set -ex        && wget https://releases.hashicorp.com/terraform/$TERRAFORM_VER/$TERRAFORM_ZIP -O /tmp/terraform_$TERRAFORM_ZIP        && unzip /tmp/terraform_$TERRAFORM_ZIP -d /usr/local/bin        && chmod 775 /usr/local/bin/terra*        && rm /tmp/terraform_$TERRAFORM_ZIP
 ---> Running in 5b4add0e0ca9
+ wget https://releases.hashicorp.com/terraform/0.6.7/terraform_"0.6.7"_linux_amd64.zip -O /tmp/terraform_terraform_"0.6.7"_linux_amd64.zip
--2015-11-27 11:40:40--  https://releases.hashicorp.com/terraform/0.6.7/terraform_%220.6.7%22_linux_amd64.zip
Resolving releases.hashicorp.com (releases.hashicorp.com)... 185.31.18.69
Connecting to releases.hashicorp.com (releases.hashicorp.com)|185.31.18.69|:443... connected.
HTTP request sent, awaiting response... 403 Forbidden
2015-11-27 11:40:40 ERROR 403: Forbidden.
```
Versions, if relevant:
```
➜  docker-terraform git:(109030254_terraform_container) docker version
Client version: 1.5.0
Client API version: 1.17
Go version (client): go1.4.2
Git commit (client): a8a31ef
OS/Arch (client): darwin/amd64
Server version: 1.5.0
Server API version: 1.17
Go version (server): go1.4.1
Git commit (server): a8a31ef
```

combor commented:
Hmm, it runs on mine docker.

```
inflating: /usr/local/bin/terraform-provisioner-chef
  inflating: /usr/local/bin/terraform-provisioner-file
  inflating: /usr/local/bin/terraform-provisioner-local-exec
  inflating: /usr/local/bin/terraform-provisioner-remote-exec
+ chmod 775 /usr/local/bin/terraform /usr/local/bin/terraform-provider-atlas /usr/local/bin/terraform-provider-aws /usr/local/bin/terraform-provider-azure /usr/local/bin/terraform-provider-cloudflare /usr/local/bin/terraform-provider-cloudstack /usr/local/bin/terraform-provider-consul /usr/local/bin/terraform-provider-digitalocean /usr/local/bin/terraform-provider-dme /usr/local/bin/terraform-provider-dnsimple /usr/local/bin/terraform-provider-docker /usr/local/bin/terraform-provider-dyn /usr/local/bin/terraform-provider-google /usr/local/bin/terraform-provider-heroku /usr/local/bin/terraform-provider-mailgun /usr/local/bin/terraform-provider-null /usr/local/bin/terraform-provider-openstack /usr/local/bin/terraform-provider-packet /usr/local/bin/terraform-provider-rundeck /usr/local/bin/terraform-provider-template /usr/local/bin/terraform-provider-terraform /usr/local/bin/terraform-provider-tls /usr/local/bin/terraform-provider-vsphere /usr/local/bin/terraform-provisioner-chef /usr/local/bin/terraform-provisioner-file /usr/local/bin/terraform-provisioner-local-exec /usr/local/bin/terraform-provisioner-remote-exec
+ rm /tmp/terraform_terraform_0.6.7_linux_amd64.zip
 ---> 66ba820d8674
Removing intermediate container 226eed7bd081
Successfully built 66ba820d8674
piotr@MacBook-Pro /tmp/x/docker-terraform (109030254_terraform_container)$ docker version
Client:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.5.1
 Git commit:   76d6bc9
 Built:        Tue Nov  3 21:18:00 UTC 2015
 OS/Arch:      darwin/amd64

Server:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   76d6bc9
 Built:        Tue Nov  3 19:20:09 UTC 2015
 OS/Arch:      linux/amd64
```

Can you try to remove quotes? 
```
< ENV TERRAFORM_ZIP terraform_$TERRAFORM_VER_linux_amd64.zip
---
> ENV TERRAFORM_ZIP terraform_"$TERRAFORM_VER"_linux_amd64.zip
```

dcarley committed:
Brace all variables

So that it's clear where a variable name begins and ends.

dcarley committed:
Remove quotes in TERRAFORM_ZIP var

This worked for @combor but was failing for me:

    Step 5 : RUN set -ex        && wget https://releases.hashicorp.com/terraform/$TERRAFORM_VER/$TERRAFORM_ZIP -O /tmp/terraform_$TERRAFORM_ZIP        && unzip /tmp/terraform_$TERRAFORM_ZIP -d /usr/local/bin        && chmod 775 /usr/local/bin/terra*        && rm /tmp/terraform_$TERRAFORM_ZIP
     ---> Running in 5b4add0e0ca9
    + wget https://releases.hashicorp.com/terraform/0.6.7/terraform_"0.6.7"_linux_amd64.zip -O /tmp/terraform_terraform_"0.6.7"_linux_amd64.zip
    --2015-11-27 11:40:40--  https://releases.hashicorp.com/terraform/0.6.7/terraform_%220.6.7%22_linux_amd64.zip
    Resolving releases.hashicorp.com (releases.hashicorp.com)... 185.31.18.69
    Connecting to releases.hashicorp.com (releases.hashicorp.com)|185.31.18.69|:443... connected.
    HTTP request sent, awaiting response... 403 Forbidden
    2015-11-27 11:40:40 ERROR 403: Forbidden.

Using this version of Docker (boot2docker):

    ➜  docker-terraform git:(109030254_terraform_container) docker version
    Client version: 1.5.0
    Client API version: 1.17
    Go version (client): go1.4.2
    Git commit (client): a8a31ef
    OS/Arch (client): darwin/amd64
    Server version: 1.5.0
    Server API version: 1.17
    Go version (server): go1.4.1
    Git commit (server): a8a31ef

This is safe to do since we're setting the variable on the preceding line
(rather than taking an input).

dcarley commented:
Yeah, that fixed it. I've pushed two commits. Could you test and let me know if you're happy with them?

combor commented:
LGTM

dcarley commented:
Cheers. Added the story ID to the subject and setup Pivotal integration for this repo.

combor commented:
Thanks. Forgot about story ID, i need to train my muscle memory :)
